### PR TITLE
chore: Refactor disaggregated serving scripts

### PR DESCRIPTION
**Pull request by pcastonguay**
To simplify disaggregated serving deployment and reduce duplicated code, the disaggregated workers and server can now be launched with:

```
mpirun -n ${NUM_RANKS} trtllm-serve disaggregated_mpi_worker -c ${CONFIG_FILE} 
trtllm-serve disaggregated -c ${CONFIG_FILE}  
```

respectively, instead of 

```
mpirun --allow-run-as-root -n ${NUM_RANKS} python3 ${EXAMPLE_DIR}/launch_disaggregated_workers.py -c ${CONFIG_FILE} 
python3 ${EXAMPLE_DIR}/launch_disaggregated_server.py -c ${CONFIG_FILE}  
```

There was some duplicated code between `launch_disaggregated_server.py` and `trtllm-serve`. So now the disaggregated server can be launched with `trtllm-serve disaggregated`.

### CODE REVIEW

**Comment by niukuo** [comment]
[PR_Github #464](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/464/) [ run ] triggered by Bot

**Comment by niukuo** [comment]
[PR_Github #468](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/468/) [ run ] triggered by Bot

**Comment by niukuo** [comment]
[PR_Github #464](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/464/) [ run ] completed with state `ABORTED`
[/LLM/main/L0_MergeRequest_PR pipeline #397](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/397/) completed with status: 'FAILURE'

**Comment by niukuo** [comment]
[PR_Github #468](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/468/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #401](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/401/) completed with status: 'FAILURE'

**Comment by niukuo** [comment]
[PR_Github #488](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/488/) [ run ] triggered by Bot

**Comment by niukuo** [comment]
[PR_Github #488](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/488/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #420](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/420/) completed with status: 'FAILURE'

File: tensorrt_llm/commands/serve.py [hunk 1]
```diff
@@ -91,68 +176,169 @@
...
+    from tensorrt_llm.llmapi.disagg_utils import split_world_comm
+
+    if (os.getenv("OMPI_COMM_WORLD_RANK")):
+        env_global_rank = int(os.environ["OMPI_COMM_WORLD_RANK"])
+    elif (os.getenv("SLURM_PROCID")):
+        env_global_rank = int(os.environ["SLURM_PROCID"])
+    else:
+        raise RuntimeError("Could not determine rank from environment")
+    device_id = env_global_rank % device_count()
+    print(
```
Comment (Tabrizian) [inline tensorrt_llm/commands/serve.py]: Can we use logger instead of print?

**Review by Tabrizian — APPROVED** [review]

**Comment by niukuo** [comment]
[PR_Github #593](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/593/) [ run ] triggered by Bot

**Comment by niukuo** [comment]
[PR_Github #593](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/593/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #503](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/503/) completed with status: 'SUCCESS'

**Comment by chuangz0** [comment]
how to run disagg with slurm  once the PR has merged?

File: tensorrt_llm/commands/serve.py [hunk 2]
```diff
@@ -91,68 +176,169 @@
...
+        env_global_rank = int(os.environ["OMPI_COMM_WORLD_RANK"])
+    elif (os.getenv("SLURM_PROCID")):
+        env_global_rank = int(os.environ["SLURM_PROCID"])
+    else:
+        raise RuntimeError("Could not determine rank from environment")
+    device_id = env_global_rank % device_count()
+    logger.info(
+        f"env_global_rank: {env_global_rank}, set device_id: {device_id} before importing mpi4py"
     )
+    torch.cuda.set_device(device_id)
```
Comment (chuangz0) [inline tensorrt_llm/commands/serve.py]: We need set device before importing mpi4py.

File: tensorrt_llm/commands/serve.py [hunk 3]
```diff
@@ -91,68 +176,169 @@
...
+@click.command("disaggregated_mpi_worker")
+@click.option("-c",
+              "--config_file",
+              type=str,
+              default=None,
+              help="Specific option for disaggregated mode.")
+def disaggregated_mpi_worker(config_file: Optional[str]):
+    """Launching disaggregated MPI worker"""
+
+    import torch
```
Comment (kaiyux) [inline tensorrt_llm/commands/serve.py]: Can we move the bunch of imports here to the head of this file?

File: examples/disaggregated/launch_disaggregated_workers.py [hunk 4]
```diff
@@ -1,130 +1,28 @@
...
-    }
-
-    llm_args = update_llm_args_with_extra_dict(llm_args, kwargs)
-
-    if backend == 'pytorch':
-        llm = PyTorchLLM(**llm_args)
+        cmd = "mpirun --allow-run-as-root"
+        cmd += ' -n ' + str(
+            total_ranks
+        ) + ' trtllm-serve disaggregated_mpi_worker -c ' + args.disagg_config_file
```
Comment (kaiyux) [inline examples/disaggregated/launch_disaggregated_workers.py]: Why do we still need this file? I'm wondering if we should directly tell users to use `trtllm-serve` in the documents, otherwise users might be asking which one they should use.

Comment (pcastonguay) [reply examples/disaggregated/launch_disaggregated_workers.py]: The advantage of using this file is that the user doesn't have to calculate the total number of MPI ranks needed for the disaggregated run, it will be computed using `get_server_configs_dict` function above. If you think that this is not needed, we can remove it, I just thought it would make things easier for the user.

Comment (kaiyux) [reply examples/disaggregated/launch_disaggregated_workers.py]: IIUC examples does not get installed after installing the pip wheel package, which is why I think all the functionality should go to `trtllm-serve` command. I see that there are more `communication_protocol` going to be supported, if those lines are going to make things easier for the users, can we port them to `trtllm-serve` command as well? Otherwise users will have to clone the repo and use it.

I don't mean to block this PR though, feel free to resolve to move faster, just thought this might not always easier the user experience.

Comment (pcastonguay) [reply examples/disaggregated/launch_disaggregated_workers.py]: I see your point. I don't have a strong objection to remove it. I've also asked @alexmsettle to provide feedback, we can try to converge when he does. Thanks.

File: tensorrt_llm/commands/serve.py (hunk 2 above)
Comment (pcastonguay) [reply tensorrt_llm/commands/serve.py]: Done.

File: examples/disaggregated/launch_disaggregated_workers.py (hunk 4 above)
Comment (alexmsettle) [reply examples/disaggregated/launch_disaggregated_workers.py]: I agree with @kaiyux there shouldn't be any code in the examples directory that is actually needed by the end users.  It should only contain example usage of code that is defined under the tensorrt_llm directory.  
So can trtllm-serve start the workers before it starts the main server?

Comment (pcastonguay) [reply examples/disaggregated/launch_disaggregated_workers.py]: In theory we could but launching multiple ctx/gen instances typically requires information about the deployment environment. For example, in a slurm cluster, we would need to launch the instances with a `srun` command such as:
```
srun -l --container-name=${container_name} \
        --container-mounts=/lustre/fsw/:/lustre/fsw,${mount_dir}:${mount_dir} \
        --mpi=pmix trtllm-serve disaggregated_mpi_worker -c disagg_config.yaml
```
but in some other clusters we might just have a regular `mpirun --hostfile <hostfile> -n 32 trtllm-serve disaggregated_mpi_worker -c disagg_config.yaml` command.
In a Kubernetes deployment, they would be created inside a pod w/o any `mpi` command, just with a `trtllm-serve` command.

So I'm leaning toward decoupling the launch of the ctx/gen instances and the disaggregated server that orchestrates the logic to send requests to ctx than to gen instances. 

I'm fine removing the `launch_disaggregated_workers.py` script since anyway right now it's designed to be used with `mpirun`, which isn't general. 

Also, in the future, we will likely want to add the ability to discover new ctx/gen instances so we could have something like:
```
#Create 1 ctx and 1 gen server
trtllm-serve TinyLlama --host localhost --port 8001
trtllm-serve TinyLlama --host localhost --port 8002

# Launch the disaggregated server with only those 2 instances
trtllm-serve disaggregated -c disagg_config.yaml
```
Then later create a new gen instance:
```
trtllm-serve TinyLlama --host localhost --port 8003
```
And use a disaggregated server API method "add_gen_instance" to add this new gen instance (not yet implemented)
```
curl http://localhost:8000/v1/add_gen_instance     -H "Content-Type: application/json"     -d '{
        "host": "localhost",
        "port": 8003,
    }' -w "\n"
```

Comment (alexmsettle) [reply examples/disaggregated/launch_disaggregated_workers.py]: I like the approach you described above - launch servers on different ports, then provide a method to add a new gen instance.

Also for starting the workers, what if instead of the script you just provide a cmd in the documenation that can be copy pasted?  Looks like each run environmen eg srun, mpi, kubernetes the main difference is the prefix before calling trtllm-serve disaggregated_mpi_worker.

```bash
mpirun --allow-run-as-root -n <num_mpi_ranks>  trtllm-serve disaggregated_mpi_worker -c disagg_config_file
```

Comment (pcastonguay) [reply examples/disaggregated/launch_disaggregated_workers.py]: > Also for starting the workers, what if instead of the script you just provide a cmd in the documenation that can be copy pasted

Yes that's what I'll do. I'll remove the script to launch the worker and document how to launch them in different cluster environments. Thanks for the feedback.

Comment (pcastonguay) [reply examples/disaggregated/launch_disaggregated_workers.py]: > I like the approach you described above - launch servers on different ports, then provide a method to add a new gen instance.

Just to be clear, even if we add that functionality, we would still want to support specifying upfront the host:port of the ctx/gen instances prior to launching the disagg server by running:
```
trtllm-serve disaggregated -c disagg_config.yaml
```
where `disagg_config.yaml` looks like:
```
hostname: localhost
port: 8000
context_servers:
  urls:
      - "localhost:8001"
      - "localhost:8002"
generation_servers:
  urls:
      - "localhost:8003"
```
And in that case, it is assumed the ctx/gen instances have already been launched as described above.

File: tensorrt_llm/commands/serve.py (hunk 3 above)
Comment (pcastonguay) [reply tensorrt_llm/commands/serve.py]: Unfortunately, I have to leave those imports here in the end. This is because according to @chuangz0 i need to set the cuda device based on MPI rank information before important mpi4py. But in the case where we run a `trtllm-serve` command without `mpi`, there's no rank information and therefore the cuda device cannot be set before importing `mpi4py`. I can only set the device based on MPI rank information if `trtllm-serve` is called with `mpirun` or `srun`. If you can think of other ways to avoid this, let me know.

**Review by juney-nvidia — APPROVED** [review]

File: examples/disaggregated/launch_disaggregated_workers.py (hunk 4 above)
Comment (alexmsettle) [reply examples/disaggregated/launch_disaggregated_workers.py]: sounds good to me

File: examples/disaggregated/README.md [hunk 5]
```diff
@@ -1,19 +1,91 @@
...
 
+## Launching context and generation servers using multiple independent `trtllm-serve` commands
+
+You can use multiple `trtllm-serve` commands to launch the context and generation servers that will be used
+for disaggregated serving. For example, you could launch two context servers and one generation servers as follows:
+
+```
+export TRTLLM_USE_UCX_KVCACHE=1
+#Context servers
+trtllm-serve TinyLlama/TinyLlama-1.1B-Chat-v1.0 --host localhsot --port 8001 --backend pytorch &> log_ctx_0 &
```
Comment (chuangz0) [inline examples/disaggregated/README.md:14]: for this case, can we specify gpus for each instance?
such as , instance 0 [tp2] use [gpu0,gpu1],instance 1 [tp4]use [gpu 2,gpu3,gpu4,gpu5]

**Review by Shixiaowei02 — APPROVED** [review]

File: examples/disaggregated/README.md (hunk 5 above)
Comment (pcastonguay) [reply examples/disaggregated/README.md:14]: You could achieve that by setting CUDA_VISIBLE_DEVICES for that `trtllm-serve` command
